### ISSUE TITLE
Implement a function for sampling and computing the log-prob

### ISSUE THREAD

**Issue by GianmarcoCallegher** [ItsNiklas/choleskyEigen#3]
Hannes wrote a code snippet for sampling and computing the log-probability of a MVN. We need to replace `jnp.linalg.cholesky` and `jnp.linalg.solve` with our code.

```python
from functools import partial

import jax
import jax.numpy as jnp
import numpy as np


@partial(jax.custom_jvp, nondiff_argnums=(0,))
def sps_mvn_sample_and_log_prob(seed, mean, inv_cov):
    sample = jax.random.normal(seed, mean.shape)
    log_prob = jnp.sum(jax.scipy.stats.norm.logpdf(sample))

    # do the following steps in c++
    ltri_inv_cov = jnp.linalg.cholesky(inv_cov)
    sample = mean + jnp.linalg.solve(ltri_inv_cov.T, sample)
    log_prob = log_prob + jnp.sum(jnp.log(jnp.diag(ltri_inv_cov)))

    return sample, log_prob


@sps_mvn_sample_and_log_prob.defjvp
def _sps_mvn_sample_and_log_prob_jvp(seed, primals, tangents):
    mean, inv_cov = primals
    mean_dot, inv_cov_dot = tangents
    sample, log_prob = sps_mvn_sample_and_log_prob(seed, mean, inv_cov)
    primals_out = (sample, log_prob)

    # do the following steps in c++

    tmp1 = sample - mean
    tmp2 = inv_cov @ tmp1

    dot1 = mean_dot @ tmp2
    dot2 = jnp.trace(jnp.linalg.solve(inv_cov, inv_cov_dot))
    dot3 = tmp1 @ inv_cov_dot @ tmp1

    tangents_out = (jnp.zeros_like(sample), dot1 + (dot2 - dot3) / 2)
    return primals_out, tangents_out


rng = np.random.default_rng(1337)

x = rng.standard_normal(10)
mean = rng.standard_normal(10)
inv_cov = np.cov(rng.standard_normal((10, 100)))


def f(seed, mean, inv_cov):
    return sps_mvn_sample_and_log_prob(seed, mean, inv_cov)[1]


g = jax.grad(f, argnums=(1, 2))

seed = jax.random.PRNGKey(1337)
print(sps_mvn_sample_and_log_prob(seed, mean, inv_cov))

## (DeviceArray([-0.48731828,  1.1614804 , -0.71950364,  0.4036346 ,
##              -0.26966274,  1.5612286 , -0.5506638 , -2.2491724 ,
##               0.84374547,  1.1257285 ], dtype=float32), DeviceArray(-15.051111, dtype=float32))

print(g(seed, mean, inv_cov))

## (DeviceArray([-1.0429684 ,  0.47583956,  0.22038332, -0.6028619 ,
##              -0.04563203, -0.04032344, -0.26514393, -1.8590978 ,
##              -0.39493063,  1.5628026 ], dtype=float32), DeviceArray([[-2.6400173 ,  0.7936772 , -0.25114876, -0.98017657,
##               -0.5188692 , -0.2423878 , -0.44342   , -2.9005187 ,
##                1.1007642 ,  2.826306  ],
##              [ 0.7936772 ,  0.34972584,  0.06823815,  0.27157778,
##                0.17574082, -0.01501111,  0.08001304,  0.84057987,
##               -0.20253313, -0.6062638 ],
##              [-0.25114876,  0.06823815,  0.5801998 , -0.08691622,
##               -0.03397314, -0.05320641, -0.01351129, -0.15539229,
##                0.00910744,  0.11294655],
##              [-0.98017657,  0.27157778, -0.08691622,  0.36904976,
##               -0.1295996 , -0.16182247, -0.21523169, -1.0346297 ,
##                0.4438629 ,  0.9537356 ],
##              [-0.5188692 ,  0.17574082, -0.03397314, -0.1295996 ,
##                0.44808584, -0.03328741, -0.02525652, -0.54740435,
##                0.20516413,  0.500587  ],
##              [-0.2423878 , -0.01501111, -0.05320641, -0.16182247,
##               -0.03328741,  0.41767797, -0.07864615, -0.2816229 ,
##                0.05100846,  0.27677017],
##              [-0.44342   ,  0.08001304, -0.01351129, -0.21523169,
##               -0.02525651, -0.07864615,  0.40243125, -0.37040278,
##                0.14975199,  0.33982256],
##              [-2.9005187 ,  0.84057987, -0.15539229, -1.0346297 ,
##               -0.54740435, -0.2816229 , -0.37040278, -2.179127  ,
##                1.0649163 ,  2.6031616 ],
##              [ 1.1007642 , -0.20253311,  0.00910743,  0.4438629 ,
##                0.20516413,  0.05100846,  0.14975199,  1.0649163 ,
##                0.01988679, -0.892138  ],
##              [ 2.826306  , -0.6062639 ,  0.11294653,  0.9537356 ,
##                0.500587  ,  0.27677017,  0.33982256,  2.6031616 ,
##               -0.892138  , -1.9402502 ]], dtype=float32))

# some basic checks

sample, _ = sps_mvn_sample_and_log_prob(seed, mean, inv_cov)


def f0(sample, mean, inv_cov):
    cov = jnp.linalg.inv(inv_cov)
    return jax.scipy.stats.multivariate_normal.logpdf(sample, mean, cov)


g0 = jax.grad(f0, argnums=(1, 2))

print(f0(sample, mean, inv_cov))

## -15.051112

print(g0(sample, mean, inv_cov))

## (DeviceArray([-1.0429688 ,  0.47583953,  0.22038335, -0.6028619 ,
##              -0.04563197, -0.04032347, -0.26514393, -1.8590977 ,
##              -0.39493063,  1.5628027 ], dtype=float32), DeviceArray([[-2.640018  ,  0.79367745, -0.25114882, -0.9801766 ,
##               -0.5188694 , -0.24238797, -0.44342   , -2.9005191 ,
##                1.1007643 ,  2.8263066 ],
##              [ 0.79367733,  0.3497258 ,  0.06823814,  0.27157775,
##                0.17574085, -0.01501109,  0.08001305,  0.8405798 ,
##               -0.20253313, -0.6062638 ],
##              [-0.25114873,  0.06823814,  0.58019984, -0.0869162 ,
##               -0.03397314, -0.0532064 , -0.01351128, -0.15539221,
##                0.00910739,  0.11294652],
##              [-0.9801768 ,  0.27157778, -0.08691623,  0.36904976,
##               -0.12959965, -0.16182253, -0.21523173, -1.0346298 ,
##                0.44386292,  0.95373577],
##              [-0.51886934,  0.17574082, -0.03397314, -0.12959959,
##                0.4480859 , -0.03328742, -0.02525649, -0.54740435,
##                0.20516413,  0.500587  ],
##              [-0.24238795, -0.01501106, -0.05320641, -0.1618225 ,
##               -0.03328743,  0.41767803, -0.07864618, -0.28162295,
##                0.0510085 ,  0.27677023],
##              [-0.44342008,  0.08001305, -0.01351129, -0.21523169,
##               -0.02525649, -0.07864617,  0.40243134, -0.37040275,
##                0.14975199,  0.33982256],
##              [-2.9005191 ,  0.8405798 , -0.15539232, -1.0346297 ,
##               -0.54740435, -0.28162295, -0.37040275, -2.1791272 ,
##                1.0649163 ,  2.6031616 ],
##              [ 1.1007643 , -0.20253314,  0.00910742,  0.4438629 ,
##                0.20516416,  0.05100849,  0.14975196,  1.0649163 ,
##                0.01988683, -0.8921381 ],
##              [ 2.8263066 , -0.606264  ,  0.11294647,  0.9537356 ,
##                0.500587  ,  0.27677023,  0.3398225 ,  2.6031613 ,
##               -0.89213794, -1.9402503 ]], dtype=float32))
```

**Comment by ItsNiklas** [comment]
JVP rule returning primals and tangents still missing.

**Comment by ItsNiklas** [comment]
Implemented in 08adfd2.
